### ISSUE TITLE
Display completion items for member variables sorted in declaration order

### ISSUE THREAD

**Issue by jeremyong** [microsoft/vscode-cpptools#9497]
### Feature Request

C++20 has released [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization) which enables the initialization of data as follows:

```cpp
struct Foo { int b; char a; };
Foo foo{ .b = 3, .a = 's' };
```

Note that the initialization of designated members _must_ be listed in declaration order (first `b`, then `a`). This is a strict error, not a warning and this behavior can only be disabled in clang (not GCC or MSVC).

The issue is that when the extension provides completion items, it will list members in alphabetical order (i.e. first `a`, then `b` in this example). This forces the user to inspect the declaration site to determine if they have the order correct (or get hit with a compiler error long after).

As for severity, this issue has a dramatic impact on usability of this feature. The example shown here is a small struct, but consider structs such as [`VkPhysicalDeviceFeatures`](https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkPhysicalDeviceFeatures.html) with dozens upon dozens of non-alphabetized fields. In production code, big bags of data or state bits such as these are common, and having declaration-order listings of completion items would be a dramatic productivity boost.

**Comment by sean-mcmanus** [comment]
It seems to me that the completion list should only show the members that are valid to use at that spot and not just sort the valid one first, right?

**Comment by jeremyong-az** [comment]
Yes that would be even better than what I've proposed. I just thought that a declaration order sort would at least be an incremental/easier change but what you've suggested sounds perfect!

**Comment by jeremyong** [comment]
Yes that would be even better than what I've proposed. I just thought that a declaration order sort would at least be an incremental/easier change but what you've suggested sounds perfect!

**Comment by sean-mcmanus** [comment]
Actually, all the members are usable for the 1st designated initializer, so sorting them is still a valid request, and the hiding of invalid ones is a separate issue.

The feature request is already being tracked by our shared VS implementation at https://developercommunity.visualstudio.com/t/Show-Intellisense-Suggested-C2a-Design/1121226 so you could upvote that.

I've filed a separate feature request for not showing the invalid completion suggestions at https://developercommunity.visualstudio.com/t/C-IntelliSense-completion-list-for-C/10079651 .

**Comment by CIOSAI** [comment]
I encountered the same issue
